### PR TITLE
[5.x] Template color scheme configuration

### PR DESCRIPTION
**Pull request by Fedik**
Pull Request for Issue #42134 .

### Summary of Changes

Add a Color scheme switch to Atum template. 
The PR allow to choose prefered color scheme based on OS setting, or fixed.
Instead of using media query (which is not possible to override) it uses data attribute (`data-bs-theme` for bootstrap).
Very similar to what @dgrammatiko tried in #42209 .

### New API
The following data attributes should be used within template:
`data-color-scheme-os` When template should follow OS setting
`data-color-scheme="light"` Use Light scheme
`data-color-scheme="dark"` Use Dark scheme

The combination `data-color-scheme-os data-color-scheme="dark"` means that the template will follow OS setting, but initialy will be used `dark` color scheme.

Addittionaly the PR introduce a new JavaScript event `joomla:color-scheme-change` which dispatched when color scheme changes.

All of this will allow to extensions easily follow to actual template state.
And in theory, should simplify Editor fixes for specified color scheme.

#### SCSS fixes needed

Following SCSS code does not work:
```scss
// Inherit parent with &
.foo{
  @if $enable-dark-mode {
    @include color-mode(dark) {
      &.bar {
        background-color: var(--template-bg-dark-90);
      }
    }
  }
}
// Child element
.foo{
  @if $enable-dark-mode {
    @include color-mode(dark) {
      .bar {
        background-color: var(--template-bg-dark-90);
      }
    }
  }
}
```
All such instance should use:
```scss
// For fisrt example
@if $enable-dark-mode {
  @include color-mode(dark) {
    .foo.bar {
      background-color: var(--template-bg-dark-90);
    }
  }
}
// For second example
@if $enable-dark-mode {
  @include color-mode(dark) {
    .foo .bar {
      background-color: var(--template-bg-dark-90);
    }
  }
}
```
(Or maybe someoen know better fixes?)

After the PR will be merged someone should look on it. Or can send PR against my PR.

### Testing Instructions
Apply patch run `npm install`.
Check the template Atum parameters `Dark mode`.
When its set to `Follow OS settings` then template color scheme should changes when OS color scheme changes.
When its set to `Use Light theme` or `Use Dark theme` then  template color scheme should stay Light or Dark, always.

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [ ] No documentation changes for docs.joomla.org needed
- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/238
- [ ] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by bembelimen** [comment]
Hey @Fedik , looks good. Do you see also an option to add the user parameters as an extra layer, so a user can define their preferences?

File: administrator/language/en-GB/tpl_atum.ini [hunk 1]
```diff
@@ -6,6 +6,10 @@
 ATUM="Atum Administrator template"
 TPL_ATUM_BACK_TO_CONTROL_PANEL="Back to Dashboard"
 TPL_ATUM_BACKEND_LOGIN="Joomla Administrator Login"
+TPL_ATUM_COLORSCHEME_LABEL="Dark mode"
```
Comment (brianteeman) [inline administrator/language/en-GB/tpl_atum.ini]:
```suggestion
TPL_ATUM_COLORSCHEME_LABEL="Dark Mode"
```

File: administrator/templates/atum/templateDetails.xml [hunk 2]
```diff
@@ -111,6 +111,16 @@
 					<option value="0">JNO</option>
 					<option value="1">JYES</option>
 				</field>
+				<field
+					name="colorScheme"
+					type="list"
+					label="TPL_ATUM_COLORSCHEME_LABEL"
+					default="0"
+				>
+					<option value="0">TPL_ATUM_COLORSCHEME_OPTION_FOLLOW_OS</option>
```
Comment (brianteeman) [inline administrator/templates/atum/templateDetails.xml]:
```suggestion
					default="0"
					>
					<option value="0">TPL_ATUM_COLORSCHEME_OPTION_FOLLOW_OS</option>
```

File: build/media_source/templates/administrator/atum/js/template.es6.js [hunk 3]
```diff
@@ -218,7 +218,36 @@ function subheadScrolling() {
...
+  const mql = window.matchMedia('(prefers-color-scheme: dark)');
+  const check = () => {
+    const newScheme = mql.matches ? 'dark' : 'light';
+    // Check if theme already was set
+    if (docEl.dataset.colorScheme === newScheme) return;
+    const expires = new Date();
+    docEl.dataset.bsTheme = newScheme;
+    docEl.dataset.colorScheme = newScheme;
+    expires.setTime(expires.getTime() + 31536000000);
+    // Store theme in cookies, so php will know the last choose
```
Comment (brianteeman) [inline build/media_source/templates/administrator/atum/js/template.es6.js]:
```suggestion
    // Store theme in cookies, so php will know the last choice
```

**Comment by Fedik** [comment]
> Do you see also an option to add the user parameters as an extra layer

I thought about it, but I would make it in separated PR.
We need a feature that will allow to define "User specific" params in templateDetails.xml. This way it will be more flexible, than hardcoding something in to User profile form.
I think some years ago there already was a discussion about it.

**Comment by HLeithner** [comment]
can't we do this in a generic way? maybe with a module and joomla w-c and event trigger and not only for atum?

**Comment by HLeithner** [comment]
also you call it color scheme switcher, which leads  to the expectation that you can more then on/off dark mode (which by the way would make sense, allowing more then just darkmode on/off) if this things is also useable for the frontend and by 3rd party

**Comment by brianteeman** [comment]
Why not do it the bootstrap way as this is for a bootstrap template?

Specifically I am redferring to the javascript so that this can be easily toggled. Currently if i understand this correctly then its a template style setting which will mean its about 5 clicks just to toggle.

https://getbootstrap.com/docs/5.3/customize/color-modes/#javascript

**Comment by Fedik** [comment]
> can't we do this in a generic way? maybe with a module and joomla w-c and event trigger and not only for atum?

You mean for User specific? Yeah, that can work also, it will be "local setting" and in each browser user should do it again. Also will need to add a Swicth somewhere in to UI.

If you mean in general, it already is as generic as possible.

> you call it color scheme switcher,

Because in Browser API it called `(prefers-color-scheme: dark)` :wink: 

> Currently if i understand this correctly then its a template style setting which will mean its about 5 clicks just to toggle.

Let's be honest, how often do you need it? I will set it once and will forget it exists.

**Comment by brianteeman** [comment]
> > Currently if i understand this correctly then its a template style setting which will mean its about 5 clicks just to toggle.

> Let's be honest, how often do you need it? I will set it once and will forget it exists.

Perhaps but perhaps not

**Comment by C-Lodder** [comment]
> Why not do it the bootstrap way as this is for a bootstrap template?

Suppose it boils down to whether Joomla wants to simply toggle between OS, light and dark, or whether they want to offer a spectrum of themes.

I'd even go as far as saying a switcher (toggle) doesn't warrant saving the preference in the database, but instead just via local storage, seeing as it's a quick and easy action.

**Comment by brianteeman** [comment]
I agree with you about using local storage. That is the most logical place to store an environmental setting as opposed to a site setting

**Comment by ceford** [comment]
With Firefox dark mode OFF and Dark Mode set to use Dark Theme the Toolbar button text has too little contrast. Otherwise OK.
screen shot 2023-10-27 at 08 35 01
With Firefox dark mode ON and Follow OS Settings selected the tab panels are stacked vertically and the Dark Mode list with Follow OS Settings or Use Light Theme selected has too little contrast - white on grey.
screen shot 2023-10-27 at 08 35 20
So: more work needed?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42221">issues.joomla.org/tracker/joomla-cms/42221</a>.</sub>

**Comment by Fedik** [comment]
> So: more work needed?

yes, this part is in the description under "SCSS fixes needed" title,
I have a hope it will be fixed in #42010, but not sure currently

**Comment by Septdir** [comment]
I don't want to offend anyone, but why should the choice of color scheme be in the template settings?

This is fundamentally the wrong approach. Same as taking your browser settings to select a theme.

Does the site only have one administrator? No, there may be a lot of them. And some people use the dark theme, and some don't.
Moreover, one person may have different color schemes on different devices.

For example, when working at a computer in good lighting, I do not use dark themes. And not about the phone, it’s very difficult for me to look at light topics.

Therefore, if you make different color schemes, then changing the scheme should be available on each page so that a person can change it at any time.

Otherwise, it is not much different from what it is now.

And once again I apologize if my words or tone offended anyone.
I just had time to look at Joomla 5 and I was simply shocked by the dark theme and the fact that I can only change it in the browser settings.

**Comment by progreccor** [comment]
I totally agree with @Septdir  . It is a smart approach.

**Comment by fiction13** [comment]
@Septdir is right!

**Comment by korenevskiy** [comment]
In the future, some day there will be settings in the browser to switch the site mode. But such times have not come yet and it will not be soon. Need a switch of 3 modes on the panel: Auto, Light, Dark. Or store the template settings mode. Or store it in the browser storage using JS.

I use the dark mode in the OS. But **MS Word** is difficult to use in dark mode, the colored text is not visible or almost invisible. How often do you switch **MS Word** to white mode to work with a document? It's the same here. It should be understood that the dark mode for viewing the site and the dark mode for editing the site are not the same thing at all. Often you need to switch just on the go. Moreover, we have to test the text in both modes so that the text is not intentionally colored by accident. A hot switch is NECESSARY.

I suggest making 2 settings, one in the template settings, the second on the panel in the profile drop-down menu or just on the panel near the Joomla spring. The first setting sets the default mode. The button on the panel will override the default mode and store the status in the browser storage.

File: administrator/components/com_users/forms/user.xml [hunk 4]
```diff
@@ -186,6 +186,19 @@
...
 
+			<field
+				name="colorScheme"
+				type="list"
+				label="COM_USERS_USER_COLORSCHEME_LABEL"
+				default=""
+				filter="options"
+			>
+				<option value="">JOPTION_USE_DEFAULT</option>
+				<option value="0">COM_USERS_USER_COLORSCHEME_OPTION_FOLLOW_OS</option>
```
Comment (Fedik) [inline administrator/components/com_users/forms/user.xml]: @bembelimen you probably wanted `value="os"` here or?

File: administrator/templates/atum/templateDetails.xml [hunk 5]
```diff
@@ -111,6 +111,17 @@
 					<option value="0">JNO</option>
 					<option value="1">JYES</option>
 				</field>
+				<field
+					name="colorScheme"
+					type="list"
+					label="TPL_ATUM_COLORSCHEME_LABEL"
+					default="os"
+					>
+					<option value="0">JDISABLED</option>
```
Comment (Fedik) [inline administrator/templates/atum/templateDetails.xml:121]: I think "Disabled" does not make sense, as it canot be disabled, it should be either one of

File: administrator/components/com_users/forms/user.xml [hunk 6]
```diff
@@ -186,6 +186,19 @@
 				<option value="">JOPTION_USE_DEFAULT</option>
 			</field>
 
+			<field
+				name="colorScheme"
+				type="list"
+				label="COM_USERS_USER_COLORSCHEME_LABEL"
+				default=""
+				filter="options"
+			>
```
Comment (brianteeman) [inline administrator/components/com_users/forms/user.xml]:
```suggestion
				>
```

File: administrator/language/en-GB/com_users.ini [hunk 7]
```diff
@@ -399,6 +399,10 @@ COM_USERS_USER_BACKUPCODES_CAPTIVE_PROMPT="If you do not have access to your usu
 COM_USERS_USER_BACKUPCODES_DESC="Lets you access the site if all other Multi-factor Authentication methods you have set up fail."
 COM_USERS_USER_BATCH_FAILED="An error was encountered while performing the batch operation: %s."
 COM_USERS_USER_BATCH_SUCCESS="Batch operation completed."
+COM_USERS_USER_COLORSCHEME_LABEL="Dark Mode"
+COM_USERS_USER_COLORSCHEME_OPTION_DARK="Use Dark theme"
```
Comment (brianteeman) [inline administrator/language/en-GB/com_users.ini]: is it correct that you refer to theme here and not mode?

File: administrator/templates/atum/templateDetails.xml [hunk 8]
```diff
@@ -111,6 +111,17 @@
 					<option value="0">JNO</option>
 					<option value="1">JYES</option>
 				</field>
+				<field
+					name="colorScheme"
+					type="list"
+					label="TPL_ATUM_COLORSCHEME_LABEL"
+					default="os"
+					>
```
Comment (brianteeman) [inline administrator/templates/atum/templateDetails.xml:120]: add validate-=options?

File: build/media_source/templates/administrator/atum/scss/_variables.scss [hunk 9]
```diff
@@ -2,7 +2,7 @@
 $prefix: "";
 
 // Use dark mode conditionally based on media queries from the browser to match the desktop theme
```
Comment (brianteeman) [inline build/media_source/templates/administrator/atum/scss/_variables.scss]: This comment needs updating now that its not a media query

File: components/com_users/forms/frontend.xml [hunk 10]
```diff
@@ -30,6 +30,19 @@
...
 				<option value="">JOPTION_USE_DEFAULT</option>
 			</field>
+
+			<field
+				name="colorScheme"
+				type="list"
+				label="COM_USERS_USER_COLORSCHEME_LABEL"
+				default=""
+				filter="options"
+			>
```
Comment (brianteeman) [inline components/com_users/forms/frontend.xml]:
```suggestion
				>
```

File: components/com_users/forms/frontend.xml [hunk 11]
```diff
@@ -30,6 +30,19 @@
 				>
 				<option value="">JOPTION_USE_DEFAULT</option>
 			</field>
+
+			<field
+				name="colorScheme"
+				type="list"
+				label="COM_USERS_USER_COLORSCHEME_LABEL"
+				default=""
+				filter="options"
```
Comment (brianteeman) [inline components/com_users/forms/frontend.xml]: validate=options ?

File: administrator/language/en-GB/com_users.ini (hunk 7 above)
Comment (Fedik) [reply administrator/language/en-GB/com_users.ini]: well, it all about color-scheme, 
probably need to update the labels

File: build/media_source/templates/administrator/atum/scss/_variables.scss (hunk 9 above)
Comment (Fedik) [reply build/media_source/templates/administrator/atum/scss/_variables.scss]: done

File: administrator/language/en-GB/com_users.ini (hunk 7 above)
Comment (Fedik) [reply administrator/language/en-GB/com_users.ini]: Changed a bit, sounds better now

**Comment by adj9** [comment]
By installing the PR from scratch the administrative homa does not have the cottect style, the titles of all boxes are to be reduced.
While the other pages , such as articles and system, work
JD

**Comment by walturbo** [comment]
@adj9 idem

home

**Comment by Fedik** [comment]
The style fix will be there:
- https://github.com/joomla/joomla-cms/pull/42757

Current PR povides only the switch feature.

**Comment by LadySolveig** [comment]
What a very nice improvement for the dark mode - thanks @Fedir for your work 🚀 🤖 and also for the support, collaboration and for testing to @bembelimen  @fgsw @brianteeman

**Comment by Quy** [comment]
@Fedik I don't remember it being this way with PR #42986. It is probably because of this PR?

42221-dark-mode

**Comment by Fedik** [comment]
Yes, it probably because of an issue in scss, I wrote in the fitrst comment.
Need to check that for fancy-select style.

**Comment by LadySolveig** [comment]
@Fedir Created a PR, sorry that we missed your comment and messed up the Fancy-Select.
Would be happy if you could give me a test @Fedik @Quy #43008

**Comment by Quy** [comment]
When installing, it is in light mode. After installing and redirecting to login page, then it is in dark mode. Expected behavior?

**Comment by Fedik** [comment]
>  After installing and redirecting to login page, then it is in dark mode.

Does your OS use Dark Mode? if yes, then it is expected I guess.

**Pull request merged**